shopfloor/services/delivery.py

```diff
@@ -206,24 +234,37 @@ def _lines_from_lot_domain(self, lot, no_qty_done=True):
             [self._lines_base_domain(no_qty_done), [("lot_id", "=", lot.id)]]
         )
 
-    def _lines_from_product_domain(self, product, no_qty_done=True):
-        return expression.AND(
+    def _lines_from_product_domain(self, product, no_qty_done=True, packaging_qty=None):
+        domain = expression.AND(
             [self._lines_base_domain(no_qty_done), [("product_id", "=", product.id)]]
         )
+        if packaging_qty:
+            # Match lines corresponding to the product packaging qty
+            # (in stock UoM) and exclude lines having a source package (those
+            # must be processed by scanning the package)
+            domain.extend(
+                [
+                    ("product_qty", ">=", packaging_qty),
+                    ("package_id", "=", False),
+                ]
+            )
+        return domain
 
     def _lines_from_package_domain(self, package, no_qty_done=True):
         return expression.AND(
             [self._lines_base_domain(no_qty_done), [("package_id", "=", package.id)]]
         )
 
-    def _deliver_product(self, picking, product):
+    def _deliver_product(self, picking, product, packaging_qty=None):
         if product.tracking in ("lot", "serial"):
             return self._response_for_deliver(
                 picking, message=self.msg_store.scan_lot_on_product_tracked_by_lot()
             )
 
+        limit = 1 if packaging_qty else None
         lines = self.env["stock.move.line"].search(
-            self._lines_from_product_domain(product)
+            self._lines_from_product_domain(product, packaging_qty=packaging_qty),
+            limit=limit,
```

Comment:
It is missing a sort by "picking scheduled date, picking id".

Reply:
This means storing the `scheduled_date` field on the move line... I'll check.

Reply:
We now store the `move_id.date` field at the line level in a new `date_planned` (filled by a migration script for upcoming releases).